FHIR/fhir-package-loader
# Initial FHIR Package Loader

**FHIR Package Loader is here!**

While this PR just shows a few changes, it should be used to review the full repository. The PR description is for the library in general and how to test it. Any change requests/comments can be coordinated on the PR but unfortunately, I think we'll need to just link to lines in the code manually since we can't comment on arbitrary lines that are not included in this PR.

This library provides separate functions for downloading FHIR packages (from `packages.fhir.org` or `packages2.fhir.org`). It can be used on it's own through a CLI, within a project by calling either `loadDependencies` (or `loadDependency` if you're SUSHI), or through an API.

A brief overview of how the library works:

- The `FHIRDefinitions` class holds all the generic functionality we would want for definitions of FHIR packages. This does not include some of the SUSHI specific ideas, like `supplementalFHIRDefinitions`. SUSHI then extends this `FHIRDefinitions` class and adds the additional properties it needs.
- Besides not including the SUSHI specific attributes, the `FHIRDefinitions` class here works a little differently. Instead of having one `FHIRDefinitions` class instance per package you download, you will receive back one `FHIRDefinitions` class instance, which then has `childFHIRDefs` for each of the packages you downloaded. This lets you better search for a definition within a specific package, supports packages that have definitions of the same name, and will hopefully better support transitive dependencies being downloaded in the future (for example, if you downloaded mCODE, this could support the library also downloading US Core because mCODE has it as a dependency). Because of this functionality, the `allProfiles`, `allExtensions`... functions support passing in a specific package name if you want to search only for definitions within a specific package.
- A path can be specified as to where the library looks to see if a package is already downloaded and downloads it to that directory if it is not already there. This defaults to the FHIR cache, which is what SUSHI uses.
- The function used to log messages is passed in to the various functions so that the function using the library decides on the method for logging and can control it's own level of logging. SUSHI passes in a function to use the FSHLogger, the CLI uses it's own winston logger, and someone else could simply pass in `console.log`.
- `loadDependencies` is used in the CLI and in the API to accept a list of packages and versions, a path to check for already downloaded packages or to download any new ones to, and a function to log any messages with, and returns a single `FHIRDefinitions` class with all definitions.
- SUSHI uses `loadDependency` because this work did not change how SUSHI works with and uses `FHIRDefinitions` as a class. Therefore, SUSHI still passes in a single `FHIRDefinitions` and the definitions are added to that instance, so they aren't separated out by package how they would be if you used the CLI or API.
- The CLI requires the packages and versions to be in the format of `package@version` so that the `#` doesn't need to be escaped on the command line. This is the same format GoFSH specifies now. The API supports either `@` or `#`.

Things to review:

- Make sure the README makes sense and describes the functionality well. It should also help you to use the library when testing things out.
- I'm not sure what the best way to do this is, but the functionality in `loadDependency` should pretty closely match up with SUSHI's old `loadDependency` function. That function did have minor changes made to it during the process of creating this library, and I think I've kept up with all of them, but there's always room for error.
- Ensure you can use the CLI
- Ensure you can use the API
- Ensure SUSHI still works (which uses the `loadDependency` function - there are instructions in the [related SUSHI PR](https://github.com/FHIR/sushi/pull/1026) on how to link the library locally)
- The tests should run and should make sense
- The repository set up should look correct (package.json fields; eslint, prettier, jest, typescript configurations; folder structure)

How to test:

- CLI
  - You can use `ts-node` to test this
  - Alternatively (or additionally), you can run `npm pack` and then `npm install -g <result-of-npm-pack>.tgz` which will show you how it will work once it's published to NPM
  - With either method, you can use the `--help` command to see how the CLI works
  - You can use the `install` command to install any packages
  - You can use the `-s` option to specific a different directly than the FHIR cache and you can use `-d` for debug log messages (FYI there are none, but you could add your own to test that it actually works lol).
  - Example: `fpl install hl7.fhir.us.mcode@1.0.0` will look to see if mCODE is installed in your FHIR cache and download it if not.
- API
  - You can import the library into a separate .js file and use the `fpl` function to access FHIR definitions. The example in the README should be sufficient (if it isn't, I should probably change it lol)
- Use one of the functions directly
  - SUSHI does this, so that might be sufficient.
  - If you do want to try it out though, you should also be able to import the library into a separate .js file and then use `loadDependencies`. You can pass in an array of packages and then get back `FHIRDefinitions` with all the package definitions.

I'm sure I forgot something here, so let me know if you have any questions about anything!
